File: src/main/java/com/flexpoker/framework/event/subscriber/InMemoryThreadSafeEventSubscriberHelper.java
package com.flexpoker.framework.event.subscriber;

import java.util.Comparator;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.PriorityBlockingQueue;

import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import com.flexpoker.framework.event.Event;
import com.flexpoker.framework.event.EventHandler;

@Scope("prototype")
@Component
public class InMemoryThreadSafeEventSubscriberHelper<T extends Event> {

    private final Map<UUID, PriorityBlockingQueue<T>> listOfEventsNeededToProcess;

    private final Map<UUID, Integer> nextExpectedEventVersion;

    private Map<Class<T>, EventHandler<T>> handlerMap;
    
    public InMemoryThreadSafeEventSubscriberHelper() {
        listOfEventsNeededToProcess = new ConcurrentHashMap<>();
        nextExpectedEventVersion = new ConcurrentHashMap<>();
    }

    public void receive(T event) {
        synchronized (this) {
            listOfEventsNeededToProcess.putIfAbsent(event.getAggregateId(), new PriorityBlockingQueue<>(10, Comparator.comparingInt(Event::getVersion)));
            nextExpectedEventVersion.putIfAbsent(event.getAggregateId(), Integer.valueOf(1));

            if (isExpectedEvent(event)) {
                handleEventAndRunAnyOthers(event);
            } else {
                listOfEventsNeededToProcess.get(event.getAggregateId()).add(event);
            }
        }
    }

    private void handleEventAndRunAnyOthers(T event) {
        handleEvent(event);
        removeEventFromUnhandleList(event);
        incrementNextEventVersion(event);
        handleAnyPreviouslyUnhandledEvents(event);
    }

    private boolean isExpectedEvent(Event event) {
        int expectedEventVersion = nextExpectedEventVersion.get(event.getAggregateId())
                .intValue();
        return expectedEventVersion == event.getVersion();
    }

    private void handleEvent(T event) {
        handlerMap.get(event.getClass()).handle(event);
    }

    private void removeEventFromUnhandleList(Event event) {
        listOfEventsNeededToProcess.get(event.getAggregateId()).remove(event);
    }

    private void incrementNextEventVersion(Event event) {
        nextExpectedEventVersion.compute(event.getAggregateId(),
                (eventId, eventVersion) -> eventVersion + 1);
    }

    private void handleAnyPreviouslyUnhandledEvents(T event) {
        T earliestUnrunTableEvent = listOfEventsNeededToProcess.get(event.getAggregateId()).peek();
        if (earliestUnrunTableEvent != null && isExpectedEvent(earliestUnrunTableEvent)) {
            handleEventAndRunAnyOthers(earliestUnrunTableEvent);
        }
    }

    public void setHandlerMap(Map<Class<T>, EventHandler<T>> handlerMap) {
        this.handlerMap = handlerMap;
    }

}

File: src/main/kotlin/com/flexpoker/framework/event/subscriber/InMemoryThreadSafeEventSubscriberHelper.kt
package com.flexpoker.framework.event.subscriber

import com.flexpoker.framework.event.Event
import com.flexpoker.framework.event.EventHandler
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component
import java.util.Comparator
import java.util.HashMap
import java.util.UUID
import java.util.concurrent.PriorityBlockingQueue

@Scope("prototype")
@Component
class InMemoryThreadSafeEventSubscriberHelper<T : Event> {

    private val listOfEventsNeededToProcess: MutableMap<UUID, PriorityBlockingQueue<T>>
    private val nextExpectedEventVersion: MutableMap<UUID, Int>
    private var handlerMap: Map<Class<T>, EventHandler<T>>? = null

    fun receive(event: T) {
        synchronized(this) {
            listOfEventsNeededToProcess.putIfAbsent(event.aggregateId,
                PriorityBlockingQueue(10, Comparator.comparingInt { it.version }))
            nextExpectedEventVersion.putIfAbsent(event.aggregateId, 1)
            if (isExpectedEvent(event)) {
                handleEventAndRunAnyOthers(event)
            } else {
                listOfEventsNeededToProcess[event.aggregateId]!!.add(event)
            }
        }
    }

    private fun handleEventAndRunAnyOthers(event: T) {
        handleEvent(event)
        removeEventFromUnhandleList(event)
        incrementNextEventVersion(event)
        handleAnyPreviouslyUnhandledEvents(event)
    }

    private fun isExpectedEvent(event: Event): Boolean {
        val expectedEventVersion = nextExpectedEventVersion[event.aggregateId]!!.toInt()
        return expectedEventVersion == event.version
    }

    private fun handleEvent(event: T) {
        handlerMap!![event.javaClass]!!.handle(event)
    }

    private fun removeEventFromUnhandleList(event: Event) {
        listOfEventsNeededToProcess[event.aggregateId]!!.remove(event)
    }

    private fun incrementNextEventVersion(event: Event) {
        nextExpectedEventVersion.compute(event.aggregateId) { _: UUID, eventVersion: Int? -> eventVersion!!.inc() }
    }

    private fun handleAnyPreviouslyUnhandledEvents(event: T) {
        val earliestUnrunTableEvent = listOfEventsNeededToProcess[event.aggregateId]!!.peek()
        if (earliestUnrunTableEvent != null && isExpectedEvent(earliestUnrunTableEvent)) {
            handleEventAndRunAnyOthers(earliestUnrunTableEvent)
        }
    }

    fun setHandlerMap(handlerMap: Map<Class<T>, EventHandler<T>>?) {
        this.handlerMap = handlerMap
    }

    init {
        listOfEventsNeededToProcess = HashMap()
        nextExpectedEventVersion = HashMap()
    }
}